### PR TITLE
Convert Harpy Singing Into a Trait

### PR DESCRIPTION
**Pull request by Mnemotechnician**
# Description
Adds a new trait that lets characters play midi without needing an instrument.

Singer a 1-point positive trait. For now it's only available for vulps and felinids as they are naturally more vocal than other species.

Also refactored harpy singing system a bit.

# TODO

- [X] Find out whatever causes this. Not critical, but probably should be fixed... 
<details><p>

image

</p></details>

<details><summary><h1>Media</h1></summary>
<p>

Outdated

image

image

</p>
</details>

# Changelog
:cl:
- add: Added a new positive trait - singer. It allows you to play midi songs without an instrument, similarly to harpies.

### CODE REVIEW

**Comment by DangerRevolution** [comment]
remove it from the trait menu

**Comment by Mnemotechnician** [comment]
> remove it from the trait menu

why so?

**Comment by DangerRevolution** [comment]
e

> > remove it from the trait menu
> 
> why so?

i updated the comment, should make more sense

**Comment by Mnemotechnician** [comment]
> Generic Trait = Only whatever instrument the Microphone is Harpy = All instruments

It only gives microphone and some second instrument I forgot

File: Content.Client/Traits/SingerSystem.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+using Content.Shared.Traits.Assorted.Systems;
```
Comment (VMSolidus) [inline Content.Client/Traits/SingerSystem.cs:1]: Nah. This isn't getting merged.

**Review by VMSolidus — CHANGES_REQUESTED** [review]

File: Content.Client/Traits/SingerSystem.cs (hunk 1 above)
Comment (Mnemotechnician) [reply Content.Client/Traits/SingerSystem.cs:1]: oops

Comment (Mnemotechnician) [reply Content.Client/Traits/SingerSystem.cs:1]: wait, what's your problem with that?

**Review by DangerRevolution — CHANGES_REQUESTED** [review]
Making the component generic is good; I have no idea how MIDI works but there should be some way to assign instruments per component. Ideally, the singer system is just the microphone instrument and then you can reuse it and add more instruments for Harpy

**Comment by Mnemotechnician** [comment]
> Making the component generic is good; I have no idea how MIDI works but there should be some way to assign instruments per component. Ideally, the singer system is just the microphone instrument and then you can reuse it and add more instruments for Harpy

image

**Comment by DangerRevolution** [comment]
> > Making the component generic is good; I have no idea how MIDI works but there should be some way to assign instruments per component. Ideally, the singer system is just the microphone instrument and then you can reuse it and add more instruments for Harpy
> 
> image

:O

File: Resources/Prototypes/Traits/skills.yml [hunk 2]
```diff
@@ -121,3 +121,25 @@
...
+    - !type:CharacterSpeciesRequirement
+      species:
+        - Vulpkanin # Naturally more vocal than humans
+        - Felinid
+  components:
+    - type: Singer
+      midiUi:
+        key: enum.InstrumentUiKey.Key
+        type: InstrumentBoundUserInterface
+    - type: Instrument
```
Comment (DangerRevolution) [inline Resources/Prototypes/Traits/skills.yml:139]:
```suggestion
    - type: Instrument
```
I only looked at the C# sorry

File: Resources/Prototypes/Traits/skills.yml [hunk 3]
```diff
@@ -121,3 +121,25 @@
...
+      allowPercussion: false
+      program: 52 # Voice
+    - type: SwappableInstrument
+      instrumentList:
+        "Voice": { 52: 0 }
+        "Acoustic": { 24: 0 } # All of them use synthesizer samples anyway...
```
Comment (DangerRevolution) [inline Resources/Prototypes/Traits/skills.yml:145]:
```suggestion
```

File: Resources/Prototypes/Traits/skills.yml [hunk 4]
```diff
@@ -121,3 +121,25 @@
       inverted: true
       species:
         - Felinid
+
+- type: trait
+  id: Singer
+  category: Auditory
+  points: -3
+  requirements:
+    - !type:CharacterSpeciesRequirement
```
Comment (DangerRevolution) [inline Resources/Prototypes/Traits/skills.yml:130]: ???????????

File: Resources/Prototypes/Traits/skills.yml [hunk 5]
```diff
@@ -121,3 +121,24 @@
       inverted: true
       species:
         - Felinid
+
+- type: trait
+  id: Singer
+  category: Auditory
+  points: -3
```
Comment (DangerRevolution) [inline Resources/Prototypes/Traits/skills.yml:128]:
```suggestion
  points: -6
```

File: Resources/Prototypes/Traits/skills.yml [hunk 6]
```diff
@@ -121,3 +121,24 @@
...
+  requirements:
```
Comment (DangerRevolution) [inline Resources/Prototypes/Traits/skills.yml:129]: Remove the species requirement

File: Resources/Prototypes/Traits/skills.yml (hunk 5 above)
Comment (VMSolidus) [inline Resources/Prototypes/Traits/skills.yml:128]:
```suggestion
  points: -1
```

File: Resources/Prototypes/Traits/skills.yml [hunk 7]
```diff
@@ -121,3 +121,24 @@
...
+    - !type:CharacterSpeciesRequirement
+      species:
+        - Vulpkanin # Naturally more vocal than humans
+        - Felinid
```
Comment (VMSolidus) [inline Resources/Prototypes/Traits/skills.yml:133]:
```suggestion
  requirements:
    - !type:CharacterSpeciesRequirement
      inverted: true
      species:
        - Harpy
```

File: Resources/Prototypes/Entities/Mobs/Species/harpy.yml [hunk 8]
```diff
@@ -140,6 +133,7 @@
     understands:
     - GalacticCommon
     - SolCommon
+
```
Comment (DangerRevolution) [inline Resources/Prototypes/Entities/Mobs/Species/harpy.yml:136]:
```suggestion
```

**Review by DangerRevolution — CHANGES_REQUESTED** [review]
cost higher, remove species requirement

**Review by VMSolidus — CHANGES_REQUESTED** [review]

File: Resources/Prototypes/Traits/skills.yml (hunk 5 above)
Comment (DangerRevolution) [reply Resources/Prototypes/Traits/skills.yml:128]: why?

**Comment by Mnemotechnician** [comment]
Fuck, uh... Turns out prototypedata class is apart of robust toolbox, I edited it locally, but the change wasn't committed... which is now a test fail.

I need to find a way to fix that...

File: Resources/Prototypes/Traits/skills.yml (hunk 5 above)
Comment (VMSolidus) [reply Resources/Prototypes/Traits/skills.yml:128]: It's an extremely minor trait.

**Comment by Mnemotechnician** [comment]
This works fine now. Singing types are now defined separately from the trait and can be re-used later, which makes things like singing parrots as easy to implement as adding `type: Singer, proto: NormalSinging` to them.

**Comment by DangerRevolution** [comment]
test fail valid

**Comment by DangerRevolution** [comment]
test fail still valid

**Comment by Mnemotechnician** [comment]
I'll look into it, but later

**Review by VMSolidus — APPROVED** [review]

**Review by DangerRevolution — APPROVED** [review]

**Pull request merged**